Native PostGIS raster data provider

elpaso opened the pull request:
## Description

This is an implementation of a PostGIS raster data provider in QGIS core.
Tiles are cached in RAM memory.

## TODO

- [x] Add support for WHERE conditions 
- [ ] ~and query layers~ (not a priority)
- [x] Add support for composite PKs (low priority)
- [x] Add support for tile PKs different than integer
- [x] Show more information in metadata properties
- [ ] ~Support endiannes when parsing WKB~ (not necessay, because PG always returns littleendian)
- [x] More tests

**Funded by: Christmas Holidays Inc.**

elpaso committed:
Test for PG raster provider

elpaso committed:
Add testdata for PG raster

elpaso committed:
Fix SRID on test data

elpaso committed:
First raster drawn!

elpaso committed:
PG raster overviews

elpaso committed:
Add more member variables

elpaso committed:
Fix SRID raster detection

elpaso committed:
Support raster without constraints

elpaso committed:
Find overviews

elpaso committed:
Make good use of overviews

elpaso committed:
Try to guess pks

elpaso committed:
Add isEmpty to qgs generic spatial index

elpaso committed:
Tiles cache for PG raster provider

elpaso committed:
Fix identify and nodata

elpaso committed:
translate some log messages

elpaso committed:
Fix tiles lifecycle and reformat code for readability

gioman commented:
Thanks Alessandro and Santa...

rouault commented:
Why not improving the GDAL PostgisRaster driver ?

elpaso commented:
> Why not improving the GDAL PostgisRaster driver ?

I tried multiple times to debug the issues without success and asked for help without receiving any: https://github.com/qgis/QGIS/issues/30392#issuecomment-561685894

Also, I'm familiar with QGIS API and not with GDAL.

There are other advantages in having it in QGIS core:
- better integration with the auth system
- re-use current low level PG connection pool implementation

rouault commented:
> without receiving any

Ah don't rely at pinging me like that. I might easily miss them. It is just lost in the noise of all QGIS notifications. I can't unfortunately read everything.

> better integration with the auth system

I guess some callback mechanism in GDAL could do the job

> re-use current low level PG connection pool implementation

As far as I remember the GDAL PG driver has a connection pool. See PostGISRasterDriver::GetConnection()

elpaso commented:
Feel free to fix the GDAL issues, complete the integration with the auth system and close this PR, I've learned some interesting things along the way and this was all volunteer work so I'll be happy whatever it is the outcome.

rouault commented:
> Feel free to fix the GDAL issues

I think I do more than my share on this :-) I'll not -1 this PR for sure. Just a bit sad seeing what I perceive a bit like effort duplication, but I can understand your perspective of feeling less familiar with GDAL code

elpaso commented:
@rouault I know you are doing a lot and I fully understand your POV.

All my initial efforts were oriented to identify the source of the issue and possibly try to fix it (in GDAL). I've spent several hours to get a full build toolchain with GDAL/PROJ from sources but when it came to debug GDAL internals and memory corruption in a multithreaded scenario I admin I was a bit lost.

rouault commented:
There might be some connection with https://github.com/qgis/QGIS/commit/9da842d48881fe54af734f033be91c2772fe1177 . Basically the strategy for the GDAL Postgisraster connection pool is to have a key where the id of the thread at creation time of the GDAL dataset is taken into account. Issues might arise if 2 datasets are created in the same thread, since they will share the same PG connection, but one is migrated to another thread afterwards. Above referenced commit is supposed to fix that (by using the same GDAL dataset, and a mutex around all calls)

elpaso committed:
Fix spelling and comments

elpaso committed:
Smarter loading of tile indexes

Indexes are now loaded on demand by extent

elpaso committed:
Fix build on travis + minor cleanup

elpaso committed:
Fix warnings and QT < 5.10 chopped

elpaso committed:
Fix some more warnings

elpaso committed:
Fix tests and pk guessing

elpaso committed:
DB manager use postgresraster provider instead of GDAL

elpaso committed:
Fast track for first tiles

elpaso committed:
PG raster: support SQL where conditions

elpaso committed:
PG fix test SELECT NULL::bool -> None

elpaso committed:
PG raster support composite PKs

elpaso committed:
PG raster: do not require db name

... in case a service file is used

elpaso committed:
PG raster performance boost by using overviews for stats

elpaso committed:
PG raster: add test to cmake

elpaso committed:
PG raster: more metadata

elpaso committed:
PG raster: fix GEOS exception when fetching tiles

elpaso committed:
PG raster: silence warning on NULL nodata value

elpaso committed:
PG raster: silence warning on different SRID

elpaso committed:
PG raster: make wkb parser more flexible

elpaso committed:
PG raster: faster band data lookup

elpaso committed:
PG raster: further speed optimization

- fix bbox index
- fetch data+metadata if index is empty (fast track for first call)
- micro-optimizations

elpaso commented:
I've tested this PR and it looks robust to me, I would have preferred a review but I'm merging it anyway because it fixes current issues with GDAL approach, we can revert if needed.

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+{
+}
+
+QgsPostgresRasterProvider::~QgsPostgresRasterProvider()
```
= default

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+
+bool QgsPostgresRasterProvider::hasSufficientPermsAndCapabilities()
+{
+  QgsDebugMsg( QStringLiteral( "Checking for permissions on the relation" ) );
```
QgsDebugMsgLevel( ...)

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+        mHasSpatialIndex = result.PQgetvalue( 0, 9 ) == 't';
+        bool ok;
+
+        mCrs = QgsCoordinateReferenceSystem::fromEpsgId( result.PQgetvalue( 0, 1 ).toLong( &ok ) );
```
I think this should be createFromSrid. But keep in mind that postgis crs handling is rather broken in QGIS and needs to be rethought for 4.0 (we should be retrieving the definition from the server corresponding to the reported postgis srs, and not using any qgis internal databases for this ... which may or may not have any match to the server's definitions.)

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+    mWidth = static_cast<long>( mExtent.width() / std::abs( mScaleX ) );
+    mIsTiled = ( mWidth != mTileWidth ) || ( mHeight != mTileHeight );
+
+    mCrs = QgsCoordinateReferenceSystem::fromEpsgId( result.PQgetvalue( 0, 9 ).toLong( &ok ) );
```
again, probably createFromSrid

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+
+  QgsPostgresProvider::Relkind kind = QgsPostgresProvider::Relkind::Unknown;
+
+  if ( type == QLatin1String( "r" ) )
```
type == 'r', etc

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+    // no primary or unique indices found
+    if ( res.PQntuples() == 0 )
+    {
+      QgsDebugMsg( QStringLiteral( "Relation has no primary key -- investigating alternatives" ) );
```
QgsDebugMsgLevel (QgsDebugMsg should only be used for errors)

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+
+      if ( type == QgsPostgresProvider::Relkind::OrdinaryTable || type == QgsPostgresProvider::Relkind::PartitionedTable )
+      {
+        QgsDebugMsg( QStringLiteral( "Relation is a table. Checking to see if it has an oid column." ) );
```
QgsDebugMsgLevel

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -0,0 +1,1600 @@
+                               QStringLiteral( "PostGIS" ), Qgis::Warning );
+  }
+
+  QgsDebugMsg( QStringLiteral( "************ STATS **************" ) );
```
Level

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.h:
```diff
@@ -0,0 +1,207 @@
+    QString mDetectedSrid;            //!< Spatial reference detected in the database
+    QString mRequestedSrid;           //!< Spatial reference requested in the uri
+    QgsPostgresConn *mConnectionRO = nullptr ; //!< Read-only database connection (initially)
+    QgsPostgresConn *mConnectionRW = nullptr ; //!< Read-write database connection (on update)
```
Do we need this? It's all read-only, right?

elpaso replied:
For now, yes. But I think editing will be possible in the future, I'd keep it even if it's currently unused.

nyalldawson commented on src/providers/postgres/raster/qgspostgresrastershareddata.cpp:
```diff
@@ -0,0 +1,397 @@
+  QgsPostgresRasterSharedData::TilesResponse result;
+
+  // First check for index existence
+  if ( mSpatialIndexes.find( request.overviewFactor ) == mSpatialIndexes.end() )
```
constFind?

elpaso replied:
there is no such method in std::map

nyalldawson reviewed (commented):
A belated review! :tongue:

elpaso commented:
@nyalldawson thank you for the review, I'll make the changes.

jgrocha commented:
@elpaso I've just compiled QGIS this Postgis raster support, but I'm not able to add Postgis rasters anymore (using the browser or DB Manager). I get the following error:
```
src/core/qgsmessagelog.cpp:29 : (logMessage) [0ms] [thread:0x561a671a9a60] 2020-01-25T17:56:10 PostGIS[2] Unsupported pixel type: 0
```
I saw your tests (great work, by the way) and everythng seems fine to me.

My rasters were added with `raster2pgsql`. Is it possible that the raster2pgsql encoding is different? I really don't know.

I'll investigate this issue further and I'll report back.

elpaso commented:
I'm happy to test your case. Just attach a DB dump.

jgrocha commented:
@elpaso I've created a small database with two tables, for the same image (with and without overviews). Both are stored in db. I've also tested the same image out-of-db.

The commends to load the raters were:
```bash
raster2pgsql -s 3763 -d -I -C -M `pwd`/carta_1990_ACEL.tif -F -t 1000x1000 public.carta_1900_gazeta | psql raster
```
```bash
raster2pgsql -s 3763 -d -I -C -M -l 4,8,16,32,64,128,256,512 `pwd`/carta_1990_ACEL.tif -F -t 1000x1000 public.carta_1900_gazeta_ovr | psql raster
```
The image is available at: http://webgis.di.uminho.pt/~jgr/QGIS/carta_1900_Gazeta.tif

The database backup is available at: http://webgis.di.uminho.pt/~jgr/QGIS/raster.backup

The database dump was create with:
```bash
pg_dump -Fc --file raster.backup raster
```

I'm really looking forward to see this working, but I'm missing something in my QGIS setup/compilation.

elpaso commented:
@jgrocha I'll have a look now, do you mind opening a bur report for the record?

elpaso commented:
@jgrocha nevermind, I've found the issue.
